src/com/amazon/ion/IonCursor.java

```diff
@@ -0,0 +1,83 @@
+package com.amazon.ion;
+
+import java.io.Closeable;
+
+/**
+ * A cursor over a stream of Ion data.
+ */
+public interface IonCursor extends Closeable {
+
+    /**
+     * Conveys the type of event that occurred as a result of operating on the cursor.
+     */
+    enum Event {
```

Comment:
Is it possible to refactor this in any way such that we can do more with the type system? E.g. Would it be possible to  have an enum of `NEEDS_DATA` and `NEEDS_INSTRUCTION` for the step in and out methods without fragmenting the API too much?

_If_ we were using Kotlin, we could do something like this:
```kotlin
sealed class Event {
    sealed interface StepInOutEvent
    sealed interface FillValueEvent
    object NeedsData: Event(), StepInOutEvent, FillValueEvent
    object NeedsInstruction: Event(), StepInOutEvent
    object ValueReady: Event(), FillValueEvent
    object StartScalar: Event()
    object StartContainer: Event()
    object EndContainer: Event()
}
```
This would allow us to specify e.g. that the `fillValue()` method returns a `FillValueEvent`.

So, what I'm asking is do you think there's a nice way to accomplish something similar with Java?

Reply:
I don't know of a clean way to achieve that in Java, but I am open to suggestions.

Reply:
I don't think there's a lot of value in defining a hierarchy of `Events`. What the events really indicate is what state the cursor is in; what you can call on it. Defining a hierarchy of `States` and returning `States` (a.k.a new `Cursors`) from cursor operations would make the invariant of each state explicit to users and avoid the need to check the validity of an operation for a given state.

I mocked it up (also in Kotlin) here: [IonCursor.kt](https://gist.github.com/rmarrowstone/03e10e2ca1b047dcfaca9f84e98a7278). What's really interesting about this is that, with immutable cursors, the user _can_ hold a stack of them and skip containers at any level. Holding on to a cursor for a container would also hold the buffer to capture any and all values within. As Cursors are GC'ed so would any data that is no longer readable.

But this is Kotlin you say! What about Java? Clean and ergonomic in JDK 8? Not so much. JDK 17 (with sealed classes and pattern matching)? Now you're talking!

Anyway this is just all food for thought. I'm not saying we should do the above _now_ let alone instead of what you're doing here.